pkg/activator/throttler_test.go

```diff
@@ -0,0 +1,249 @@
+			endpointsGetter: nonExistingEndpointsGetter,
+		},
+	}
+	for _, s := range samples {
+		t.Run(s.label, func(t *testing.T) {
+			var got int32
+			var received string
+			want := s.wantBreakers
+			throttler := getThrottler(defaultMaxConcurrency, s.revisionGetter, s.endpointsGetter, t)
+			if s.addCapacity {
+				throttler.UpdateCapacity(revID, 1)
+			}
+			err := throttler.Try(revID, func() {
+				got++
+			})
+			if s.wantError != "" {
+				received = err.Error()
+			}
+			if received != s.wantError {
+				t.Errorf("Expected error in the Try. Want %s, got %s", s.wantError, received)
+			}
+			if got != want {
+				t.Errorf("Unexpected number of function runs in Try. Want %d, got %d", want, got)
+			}
+		})
+	}
+}
+
+func TestThrottler_Remove(t *testing.T) {
+	throttler := getThrottler(defaultMaxConcurrency, existingRevisionGetter(10), existingEndpointsGetter, t)
+	throttler.breakers[revID] = queue.NewBreaker(throttler.breakerParams)
+	got := len(throttler.breakers)
+	if got != 1 {
+		t.Errorf("Unexpected number of Breakers was created. Want %d, got %d", 1, got)
+	}
+	throttler.Remove(revID)
+	got = len(throttler.breakers)
+	if got != 0 {
+		t.Errorf("Unexpected number of Breakers was created. Want %d, got %d", 0, got)
+	}
+}
+
+func TestHelper_UpdateEndpoints(t *testing.T) {
+	want := int32(10)
+	throttler := getThrottler(defaultMaxConcurrency, existingRevisionGetter(10), existingEndpointsGetter, t)
+	throttler.breakers[revID] = queue.NewBreaker(throttler.breakerParams)
+	updater := UpdateEndpoints(throttler)
+
+	endpointsBefore := corev1.Endpoints{ObjectMeta: v1.ObjectMeta{Name: revID.Name + "-service", Namespace: revID.Namespace}, Subsets: testinghelper.GetTestEndpointsSubset(0, 1)}
+	endpointsAfter := corev1.Endpoints{ObjectMeta: v1.ObjectMeta{Name: revID.Name + "-service", Namespace: revID.Namespace}, Subsets: testinghelper.GetTestEndpointsSubset(1, 1)}
+	updater(&endpointsBefore, &endpointsAfter)
+
+	breaker, _ := throttler.breakers[revID]
+	got := breaker.Capacity()
+	if got != want {
+		t.Errorf("Unexpected Breaker capacity received. Want %d, got %d", want, got)
+	}
+}
+
+func TestHelper_UpdateEndpoints_WithDeltaMoreThenOne(t *testing.T) {
+	want := int32(20)
+	throttler := getThrottler(int32(20), existingRevisionGetter(10), existingEndpointsGetter, t)
+	throttler.breakers[revID] = queue.NewBreaker(throttler.breakerParams)
+	updater := UpdateEndpoints(throttler)
+
+	endpointsBefore := corev1.Endpoints{ObjectMeta: v1.ObjectMeta{Name: revID.Name + "-service", Namespace: revID.Namespace}, Subsets: testinghelper.GetTestEndpointsSubset(0, 1)}
+	endpointsAfter := corev1.Endpoints{ObjectMeta: v1.ObjectMeta{Name: revID.Name + "-service", Namespace: revID.Namespace}, Subsets: testinghelper.GetTestEndpointsSubset(2, 1)}
+	updater(&endpointsBefore, &endpointsAfter)
+
+	breaker, _ := throttler.breakers[revID]
+	got := breaker.Capacity()
+	if got != want {
+		t.Errorf("Unexpected Breaker capacity received. Want %d, got %d", want, got)
+	}
+}
```

Comment:
Should we make the "UpdateEndpoints" tests table tests as well and test some more cases? Like "delta=0" or a negative delta?

I'd envision a struct like:

```golang
{
  name string,
  concurrency int,
  endpointsBefore int,
  endpointsAfter int,
  wantCapacity int
}
```

Reply:
Yes, will add those cases